src/lib/esnext.weakref.d.ts

```diff
@@ -0,0 +1,67 @@
+interface WeakRef<T extends object> {
+    readonly [Symbol.toStringTag]: "WeakRef";
+
+    /**
+     * Returns the WeakRef instance's target object, or undefined if the target object has been
+     * reclaimed.
+     */
+    deref(): T | undefined;
+}
+
+interface WeakRefConstructor {
+    readonly prototype: WeakRef<any>;
+
+    /**
+     * Creates a WeakRef instance for the given target object.
+     * @param target The target object for the WeakRef instance.
+     */
+    new<T extends object>(target?: T): WeakRef<T>;
+}
+
+declare var WeakRef: WeakRefConstructor;
+
+interface FinalizationRegistry {
+    readonly [Symbol.toStringTag]: "FinalizationRegistry";
+
+    /**
+     * Registers an object with the registry.
+     * @param target The target object to register.
+     * @param heldValue The value to pass to the finalizer for this object. This cannot be the
+     * target object.
+     * @param unregisterToken The token to pass to the unregister method to unregister the target
+     * object. If provided (and not undefined), this must be an object. If not provided, the target
+     * cannot be unregistered.
+     */
+    register(target: object, heldValue: any, unregisterToken?: object): void;
+
+    /**
+     * Unregisters an object from the registry.
+     * @param unregisterToken The token that was used as the unregisterToken argument when calling
+     * register to register the target object.
+     */
+    unregister(unregisterToken: object): void;
+
+    /**
+     * Triggers callbacks for an implementation-chosen number of objects in the registry that have
+     * been reclaimed but whose callbacks have not yet been called.
+     *
+     * Note: The number of entries for reclaimed objects that are cleaned up from the registry
+     * (calling the cleanup callbacks) is implementation-defined. An implementation might remove
+     * just one eligible entry, or all eligible entries, or somewhere in between.
+     * @param callback If given, the registry uses the given callback instead of the one it was
+     * created with.
+     */
+    cleanupSome(callback?: (heldValue: any) => void): void;

```

Comment:
it's weird that `heldValue: any` but `target: object` and `unregisterToken: object`. I know the proposal uses the terms 'value' and 'object' but I don't think they exactly match the type system's definitions. In particular, `object` means "non-primitive". Can you provide `null`, `undefined`, `"foo"`, `false` or `3` for `target` or `unregisterToken`?

Reply:
1. `target` must be an object, i.e. non-primitive type.
   Specification: https://tc39.es/proposal-weakrefs/#sec-weak-ref-target 
   V8 implementation behavior:
     
    ```
    $ node --harmony-weak-refs
    Welcome to Node.js v14.0.0.
    Type ".help" for more information.
    > new WeakRef(null)
    Uncaught TypeError: WeakRef: target must be an object
        at new WeakRef (<anonymous>)
    > new WeakRef(undefined)
    Uncaught TypeError: WeakRef: target must be an object
        at new WeakRef (<anonymous>)
    > new WeakRef(1)
    Uncaught TypeError: WeakRef: target must be an object
        at new WeakRef (<anonymous>)
    > new WeakRef("str")
    Uncaught TypeError: WeakRef: target must be an object
        at new WeakRef (<anonymous>)
    ```

2. `unregisterToken` must be an object (non-primitive type), or undefined.

    Specification: https://tc39.es/proposal-weakrefs/#sec-finalization-registry.prototype.register
    V8 implementation behavior (please note that v8 in Node.js still uses old specification name FinalizationGroup instead of FinalizationRegistry):
     
    ```
    $ node --harmony-weak-refs
    > var fg = new FinalizationGroup(() => {})
    undefined
    > fg.register({}, null, null)
    Uncaught TypeError: unregisterToken ('null') must be an object
        at FinalizationGroup.register (<anonymous>)
    > fg.register({}, null, undefined)
    undefined
    > fg.register({}, null, 1)
    Uncaught TypeError: unregisterToken ('1') must be an object
        at FinalizationGroup.register (<anonymous>)
    > fg.register({}, null, 'a')
    Uncaught TypeError: unregisterToken ('a') must be an object
        at FinalizationGroup.register (<anonymous>)
    > fg.register({}, null, false)
    Uncaught TypeError: unregisterToken ('false') must be an object
        at FinalizationGroup.register (<anonymous>)
    ```

3. `heldValue` can have any type. 

    Specification: https://tc39.es/proposal-weakrefs/#sec-finalization-registry.prototype.register
    V8 implementation behavior:
    ```
    $ node --harmony-weak-refs
    > var fg = new FinalizationGroup(() => {})
    undefined
    > fg.register({}, null)
    undefined
    > fg.register({}, false)
    undefined
    > fg.register({}, 1)
    > fg.register({}, {})
    undefined
    ```

Reply:
I'd suggest omitting cleanupSome from .d.ts, or otherwise somehow mark it as dispreferred/not tab-complete it. It's "normative optional" in the specification, it's currently under some debate, and it's not going to be part of what V8/Chrome ships initially. See https://github.com/whatwg/html/issues/5446 for context.

Reply:
If it's "normative optional" in the specification, it can be marked as optional in .d.ts. Anyway this declaration is available only under `--lib esnext` compiler flag, which means that it's not stable. If in final stage of the proposal this method will be removed, corresponding TypeScript declaration should reflect this change before move to stable `es2020` lib.

Reply:
If Chrome's not going to ship with it initially, let's make this optional to start so that people have to prove that it's available before using it. (Only when `strict: true` of course.)

```suggestion
    cleanupSome?(callback?: (heldValue: any) => void): void;
```